### PR TITLE
Add `Disabled` field for `AdmissionPlugin` in `ShootSpec`

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area usability
/kind enhancement

**What this PR does / why we need it**:
This PR adds the option to disable an `AdmissionPlugin` in `spec.kubernetes.kubeAPIServer`.
Users have to migrate their PSPs to `PodSecurity` and disable the `PodSecurityPolicy` admission plugin using this field before upgrading to kubernetes `v1.25`.

Note: `PodSecurityPolicy` is added to requiredPlugins because we need it till all the extensions are adapted and released

**Which issue(s) this PR fixes**:
Part of #5250 

**Special notes for your reviewer**:

**Release note**:

```other operator
It is now possible to disable an admission plugin for the shoot kube-apiserver in the `ShootSpec` by setting the AdmissionPlugin.Disabled field to `true`.
```

### CODE REVIEW

File: example/90-shoot.yaml [hunk 1]
```diff
@@ -161,6 +161,7 @@ spec:
   #         clusterDefaultNodeSelector: <node-selectors-labels>
   #         namespace1: <node-selectors-labels>
   #         namespace2: <node-selectors-labels>
+  #     disabled: false
```
Comment (rfranzke) [inline example/90-shoot.yaml]: Nit: Move this before `config` after `name`

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 2]
```diff
@@ -320,6 +319,7 @@ var _ = Describe("KubeAPIServer", func() {
 						{Name: "StorageObjectInUseProtection"},
 						{Name: "MutatingAdmissionWebhook"},
 						{Name: "ValidatingAdmissionWebhook"},
+						{Name: "NamespaceLifecycle", Config: &runtime.RawExtension{Raw: []byte("namespace-lifecycle-config")}},
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go]: Why was this moved? I think the order matters 👀

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 3]
```diff
@@ -331,7 +331,31 @@ var _ = Describe("KubeAPIServer", func() {
...
+						{Name: "StorageObjectInUseProtection"},
+						{Name: "MutatingAdmissionWebhook"},
+						{Name: "ValidatingAdmissionWebhook"},
+						{Name: "NamespaceLifecycle", Config: &runtime.RawExtension{Raw: []byte("namespace-lifecycle-config")}},
+						{Name: "Foo"},
+						{Name: "Bar"},
+						{Name: "Baz", Config: &runtime.RawExtension{Raw: []byte("baz-config")}},
+					},
+				),
+				Entry("default plugins with overrides and skipping  configured plugins if disabled",
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go]:
```suggestion
				Entry("default plugins with overrides and skipping configured plugins if disabled",
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
Can you add some validation to make sure that the essentials of our default plugins cannot be disabled (https://github.com/gardener/gardener/blob/5b3456739b23d2009b69e65b642892c507842a6a/pkg/utils/kubernetes/admissionplugins.go#L26-L39)?

I suggest to forbid disabling

- Priority
- NamespaceLifecycle
- NodeRestriction
- StorageObjectInUseProtection
- MutatingAdmissionWebhook
- ValidatingAdmissionWebhook

since it can interfere with Gardener operations.

File: pkg/apis/core/types_shoot.go [hunk 4]
```diff
@@ -627,6 +627,8 @@ type OpenIDConnectClientAuthentication struct {
 type AdmissionPlugin struct {
 	// Name is the name of the plugin.
 	Name string
+	// Disabled describes whether this plugin should be disabled in the kube-apiserver
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go]: Nit:

```suggestion
	// Disabled specifies whether this plugin should be disabled.
```

Same for other types files.

File: pkg/apis/core/validation/shoot.go [hunk 5]
```diff
@@ -765,6 +773,10 @@ func validateKubernetes(kubernetes core.Kubernetes, dockerConfigured, shootHasDe
 			if forbiddenAdmissionPlugins.Has(plugin.Name) {
 				allErrs = append(allErrs, field.Forbidden(idxPath.Child("name"), fmt.Sprintf("forbidden admission plugin was specified - do not use %+v", forbiddenAdmissionPlugins.UnsortedList())))
 			}
+
+			if pointer.BoolDeref(plugin.Disabled, false) && requiredAdmissionPlugins.Has(plugin.Name) {
+				allErrs = append(allErrs, field.Forbidden(idxPath, fmt.Sprintf("the admission plugin: %s is required for Gardener operations, do not disable it", plugin.Name)))
```
Comment (rfranzke) [inline pkg/apis/core/validation/shoot.go]:
```suggestion
				allErrs = append(allErrs, field.Forbidden(idxPath, fmt.Sprintf("admission plugin %q cannot be disabled", plugin.Name)))
```

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve

**Comment by shafeeqes** [comment]
@rfranzke should we add `PodSecurityPolicy` to requiredPlugins till all the extensions are adapted and released?

File: extensions/pkg/util/shoot_test.go [hunk 6]
```diff
@@ -71,4 +73,43 @@ var _ = Describe("Shoot", func() {
...
+	Describe("IsPSPDisabled", func() {
+		var shoot = &gardencorev1beta1.Shoot{
+			Spec: gardencorev1beta1.ShootSpec{
+				Kubernetes: gardencorev1beta1.Kubernetes{
+					KubeAPIServer: &gardencorev1beta1.KubeAPIServerConfig{},
+				},
+			},
+		}
+
+		It("should return true if PodSecurityPolicy admissionPlugin is disabled", func() {
```
Comment (acumino) [inline extensions/pkg/util/shoot_test.go:86]: Can you please add a test if `kubeapiserver` is nil?

File: extensions/pkg/util/shoot.go [hunk 7]
```diff
@@ -42,3 +45,15 @@ func VersionInfo(vs string) (*version.Info, error) {
 		GitVersion: fmt.Sprintf("v%d.%d.%d", v.Major(), v.Minor(), v.Patch()),
 	}, nil
 }
+
+// IsPSPDisabled returns true if the PodSecurityPolicy plugin is explicitly disabled in the ShootSpec
+func IsPSPDisabled(shoot *gardencorev1beta1.Shoot) bool {
+	if shoot.Spec.Kubernetes.KubeAPIServer != nil {
+		for _, plugin := range shoot.Spec.Kubernetes.KubeAPIServer.AdmissionPlugins {
+			if plugin.Name == "PodSecurityPolicy" && pointer.BoolDeref(plugin.Disabled, false) {
```
Comment (acumino) [inline extensions/pkg/util/shoot.go:53]: Should we use some constant for `PodSecurityPolicy `?

File: extensions/pkg/util/shoot_test.go [hunk 8]
```diff
@@ -71,4 +73,43 @@ var _ = Describe("Shoot", func() {
 			Expect(err).NotTo(HaveOccurred())
 		})
 	})
+
+	Describe("IsPSPDisabled", func() {
```
Comment (acumino) [inline extensions/pkg/util/shoot_test.go]: Nit
```suggestion
	Describe("#IsPSPDisabled", func() {
```

File: extensions/pkg/util/shoot_test.go (hunk 6 above)
Comment (shafeeqes) [reply extensions/pkg/util/shoot_test.go:86]: Why?

File: extensions/pkg/util/shoot.go (hunk 7 above)
Comment (shafeeqes) [reply extensions/pkg/util/shoot.go:53]: Hmm, I vote against it.

File: extensions/pkg/util/shoot_test.go (hunk 6 above)
Comment (acumino) [reply extensions/pkg/util/shoot_test.go:86]: You have put a check if kubeapiserver is not nil in the helper function. So we should have a test case where if kubeapiserver is nil the it should return true.

Comment (shafeeqes) [reply extensions/pkg/util/shoot_test.go:86]: it returns false, right? I feel that's not what this function is focused on.

**Review by rfranzke — COMMENTED** [review]
/lgtm
